common/parallel/TreeCrawler.go

```diff
@@ -90,25 +127,228 @@ func (c *crawler) runWorkersToCompletion(ctx context.Context) {
 	wg := &sync.WaitGroup{}
 	for i := 0; i < c.parallelism; i++ {
 		wg.Add(1)
-		go c.workerLoop(ctx, wg, i)
+		if c.tqueue != nil {
+			go c.workerLoop(ctx, wg, i, true)
+		} else {
+			go c.workerLoop(ctx, wg, i, false)
+		}
 	}
 	wg.Wait()
 }
 
-func (c *crawler) workerLoop(ctx context.Context, wg *sync.WaitGroup, workerIndex int) {
+func (c *crawler) workerLoop(ctx context.Context, wg *sync.WaitGroup, workerIndex int, autoTune bool) {
 	defer wg.Done()
 
 	var err error
 	mayHaveMore := true
 	for mayHaveMore && ctx.Err() == nil {
-		mayHaveMore, err = c.processOneDirectory(ctx, workerIndex)
+		if autoTune {
+			mayHaveMore, err = c.processOneDirectoryWithAutoPacer(ctx, workerIndex)
+		} else {
+			mayHaveMore, err = c.processOneDirectory(ctx, workerIndex)
+		}
 		if err != nil {
 			c.output <- CrawlResult{err: err}
 			// output the error, but we don't necessarily stop the enumeration (e.g. it might be one unreadable dir)
 		}
 	}
 }
 
+func (c *crawler) receiveTqueue(root Directory) {
+
+	for ch := range c.tqueue {
+		c.cond.L.Lock()
+
+		if ch != nil {
+			c.unstartedDirs = append(c.unstartedDirs, ch)
+		}
+		c.cond.L.Unlock()
+		c.cond.Broadcast()
+	}
+
+	c.cond.L.Lock()
+	c.tqueueClosed = true
+	c.cond.L.Unlock()
+	c.cond.Broadcast()
+}
+
+// TODO: Need to pass it parallel crawl.
+const SizeInGB = 1024 * 1024 * 1024
+
+func (c *crawler) autoPacerWait(ctx context.Context) {
+	// Return in case its not sync operation.
+	if !c.isSync {
+		return
+	}
+
+	// Destination case where we return if tqueue channel closed or if some directories need processing.
+	// TODO: It has one drawback if destination is faster than source it will lead to closing of go-routine threads which we don't want.
+	if c.isSync && !c.isSource {
+		for {
+			c.cond.L.Lock()
+			if len(c.unstartedDirs) == 0 && !c.tqueueClosed && ctx.Err() == nil {
+				c.cond.L.Unlock()
+				time.Sleep(100 * time.Millisecond)
+				continue
+			}
+			c.cond.L.Unlock()
+			return
+		}
+	}
```

Comment:
Target traverser should not be slowed down by autopacer.
It'll be controlled by the rate at which directories are added to tqueue by the source traverser, so anytime target traverser waits is when tqueue channel is empty and never does an auto-wait.

Reply:
tqueue channel not the right gurantee ,
1. tqueue can be empty as source traverser doing folder enumeration.
2. tqueue can be empty most of time as soon source will put we removed from it and add to target c.unstarted dirs. 
3. tqueue not empty but go routine need to remove it, in meantime processOneDir function will be called. 
So i believe target slice which hold directories need to be enumerated will be the right check. 
We should name tqueue to something else as tqueue here is communication channel not exactly queue. unstartedDirs is target queue.